chapters/optimization/index.md

```diff
@@ -0,0 +1,145 @@
+
+```yaml
+!Flow
+version: '1'
+env:
+  JINA_ENCODER_LAYER_VAR: ${{JINA_ENCODER_LAYER}}
+pods:
+  - uses: encoder.yml
+  - uses: EuclideanEvaluator
+```
+
+The `EuclideanEvaluator` is used for calculating the distance between the encoded and an expected vector.
+Furthermore, we need the corresponding `encoder.yml`:
+
+```yaml
+!SimpleEncoder
+with:
+  index: ${{JINA_ENCODER_LAYER_VAR}}
+```
+
+```python
+class SimpleEncoder(BaseEncoder):
+
+    ENCODE_LOOKUP = {
+        'cat': [1, 3, 5],
+        'pokemon': [2, 4, 7],
+        'dog': [0, 2, 5],
+    }
+
+    def __init__(self, layer=0, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._layer = layer
+
+    def encode(self, data: 'np.ndarray', *args, **kwargs) -> 'np.ndarray':
+        return np.array([[self.ENCODE_LOOKUP[data[0]][self._layer]]])
+```
+
+The `SimpleEncoder` is not doing any computation.
+It rather illustrates the output a real encoder would have on different layers (the second layer for 'pokemon' would result in the encoding `[4]`).
+
+As the next step we need some ground truth data.
+
+```python
+documents = [
+    (Document(content='cat'), Document(embedding=np.array([2]))),
+    (Document(content='pokemon'), Document(embedding=np.array([3]))),
+    (Document(content='dog'), Document(embedding=np.array([3])))
+]
+```
+
+Documents will be send in pairs `(doc, groundtruth)` to the Flow.
+The _doc_ represents a Document that should be encoded.
+The _groundtruth_ contains the ideal encoding.
+The perfect semantic encoding for `cat` would be `2`.
+
+*Note*: In a real world example the groundtruth would rather be documents, that should be retrieved after querying.
+For the sake of simplicity we omitted the indexing step in this example.
+
+The _FlowRunner_ wraps the Flow and the Documents for rerunnability.
+This ensures no side effects between different Flow runs during optimization.
+
+```python
+runner = SingleFlowRunner('flow.yml', documents, 1, 'search', overwrite_workspace=True)
+```
+
+Now we need to tell the optimizer, what it can optimize:
+The `JINA_ENCODER_LAYER` variable.
+This is done via a `parameter.yml` file:
+
+```yaml
+- !IntegerParameter
+  jaml_variable: JINA_ENCODER_LAYER
+  high: 2
+  low: 0
+  step_size: 1
+```
+
+Here we say, the variable can take `int` values in the range `[0, 2]`.
+Jina leverages the optuna optimizer under the hood.
+For a detailed description of all possible choices see (TODO: INSERT LINK TO PARAMETER DOCSTRING).
+
+Finally, we can define the FlowOptimizer and run it:
+
+```python
+optimizer = FlowOptimizer(
+    flow_runner=runner,
+    parameter_yaml='parameter.yml',
+    evaluation_callback=MeanEvaluationCallback(),
+    n_trials=3,
+    direction='minimize',
+    seed=1
+)
+result = optimizer.optimize_flow()
+```
+
+The `MeanEvaluationCallback` takes the results of the last Evaluator inside a Flow and averages the results.
+In the above defined Flow it is the solemn `EuclideanEvaluator`.
```

Comment:
Solemn?

Reply:
My english... I thought it means something different :D